les/fetcher.go

```diff
@@ -381,12 +388,29 @@ func (f *lightFetcher) mainloop() {
 			reset(ev.Block.Header())
 			number := localHead.Number.Uint64()
 
-			// Clean stale announcements from trusted server.
-			if ulc {
-				f.forEachPeer(func(id enode.ID, p *fetcherPeer) bool {
-					p.forwardTrustedAnno(number)
-					return true
-				})
+			// Clean stale announcements from les-servers.
+			var droplist []enode.ID
+			f.forEachPeer(func(id enode.ID, p *fetcherPeer) bool {
+				removed := p.forwardAnno(number)
+				for _, anno := range removed {
+					if header := f.chain.GetHeaderByHash(anno.data.Hash); header != nil {
```

Comment:
At first I thought that this check is going to be simpler and we just need to add the check after delivering a header to the block fetcher by `f.fetcher.FilterHeaders`. Unfortunately that function returns before actually inserting the header to the chain so we still don't have the Td calculated. I guess this is why you are checking after evicting the announcements from the queue. This might work too but then I think you should also do this check in the other case when you are evicting the announcement because the queue is full (this is actually the more likely case when an attacker is spamming with bad announcements). Then we can be sure that causing the client to request useless headers by fake announcements is always punished. Either the hash is non-existent (in which case the retrieval will fail) or the hash exists but the number/Td does not match (in which case this check is always going to catch it when it gets out of the queue one way or the other).

Reply:
Think about it. We have two cases to evict announces: (a) local chain has inserted some headers, to evict stale or useless announces (b) the announce queue is full.

In the latter one, we can't do any meaningful check since they are all "future" announces.

Reply:
This is not necessarily true if the server is lying about the Td (which is the primary attack vector). If the headers exist but it is a worthless sidechain (or a fork like ETC) and the attacking server just keeps feeding announcements of it with fake high Tds then it is never evicted by `forwardAnno` since the Td appearing in the FIFO list is high (higher than the real chain). It will be evicted by `addAnno` where there is no check.

Reply:
Actually it's still ok. Image the client keeps feeding super high td all the time, it will finally trigger a syncing. And the downloader will detect this announcement is invalid and drop this peer.

Reply:
And also it's very hard to differentiate "fake announcement with high td" and "valid announcement because local client is out of sync". The main difference is for the latter one, we can finally sync to this point, but for the former, we will never reach the announcement point(it can be dropped by the downloader)